File: Java/src/main/java/com/DSA/Java/Graphs/TopologicalSort/TpsUsingMatrix.java
package com.DSA.Java.Graphs.TopologicalSort;
import java.util.ArrayList;
import java.util.Stack;

import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyMatrix.*;;

public class TpsUsingMatrix extends Graph {
    private void tpsHelper(GraphNode node, Stack<GraphNode> stack) {
        ArrayList<GraphNode> neighbours = getNeighbours(node);
        for (GraphNode tempNode : neighbours) {
            if (!tempNode.isVisited) {
                tpsHelper(tempNode, stack);
            }
        }

        node.isVisited = true;
        stack.push(node);
    }

    public void tps() {
        Stack<GraphNode> stack = new Stack<>();
        for (GraphNode node : getNodeList()) {
            if (!node.isVisited) {
                tpsHelper(node, stack);
            }
        }

        while (!stack.isEmpty()) {
            System.out.print(stack.pop().name + " ");
        }
        System.out.println();
    }
}

File: C++/Graph/src/TopologicalSort/TpsUsingMatrix.cpp
#include "../../include/TopologicalSort/TpsUsingMatrix.h"

TpsUsingMatrix::TpsUsingMatrix() {
  AdjacencyMatrix();
}

void TpsUsingMatrix::tpsHelper(GraphNode *node, std::stack<GraphNode *> &stack) {
  std::vector<GraphNode *> neighbours = get_neighbours(node, get_graph(), get_node_list());
  for (const auto &neighbour : neighbours) {
    if (!neighbour->is_visited) {
      tpsHelper(neighbour, stack);
    }
  }

  node->is_visited = true;
  stack.push(node);
}

void TpsUsingMatrix::tps() {
  std::vector<GraphNode *> nodes = get_node_list();
  std::stack<GraphNode *> stack;

  for (const auto &node : nodes) {
    if (!node->is_visited) {
      tpsHelper(node, stack);
    }
  }

  std::cout << "\n";
  while (!stack.empty()) {
    GraphNode *current_node = stack.top();
    stack.pop();
    std::cout << current_node->name << " ";
  }
  std::cout << "\n";
}
